research/picai/prepare_data.py

```diff
@@ -0,0 +1,209 @@
+                tup
+                for tup in scan_annotation_label_list
+                if any([subject_id in str(tup[1]) for subject_id in split["val"]])
+            ]
+            train_d["image_paths"], train_d["label_paths"], train_d["case_label"] = zip(*train_list)
+            val_d["image_paths"], val_d["label_paths"], val_d["case_label"] = zip(*val_list)
+
+            # Create path of json files and write to disk
+            train_write_path = os.path.join(write_dir, f"train-fold-{i}.json")
+            val_write_path = os.path.join(write_dir, f"val-fold-{i}.json")
+
+            with open(train_write_path, "w") as f:
+                json.dump(train_d, f)
+
+            with open(val_write_path, "w") as f:
+                json.dump(val_d, f)
+
+
+def preprare_data(
+    scans_read_dir: Path,
+    annotation_read_dir: Path,
+    scans_write_dir: Path,
+    annotation_write_dir: Path,
+    overview_write_dir: Path,
+    size: Optional[Sequence[int]] = None,
+    physical_size: Optional[Sequence[int]] = None,
+    spacing: Optional[Sequence[int]] = None,
+    scan_extension: str = "mha",
+    annotation_extension: str = ".nii.gz",
+    num_threads: int = 4,
+    splits_path: Optional[Path] = None,
+) -> None:
+    settings = PreprocessingSettings(
+        scans_write_dir,
+        annotation_write_dir,
+        size,
+        physical_size,
+        spacing,
+    )
+    valid_annotation_filenames = [f for f in os.listdir(annotation_read_dir) if f.endswith(annotation_extension)]
+
+    samples = []
+    for annotation_filename in valid_annotation_filenames:
+        # Annotation filename is subject id (ie patient_id study id)
+        # We use it to get the corresponding scan paths
+        annotation_path = Path(os.path.join(annotation_read_dir, annotation_filename))
+        annotation_base_filename = annotation_filename.split(".")[0]
+
+        # All of the scans for a given patient, even if there are multiple sets from different study ids,
+        # are stored in a single folder under patient id in the raw data.
+        patient_id, study_id = annotation_base_filename.split("_")
+        scans_path = os.path.join(scans_read_dir, patient_id)
+
+        # Get all the filenames that match the extension we expect and have the proper study id.
+        scan_filenames = [
+            path for path in os.listdir(scans_path) if study_id in path and path.endswith(scan_extension)
+        ]
+
+        # Extract the scans that we care about (t2w, adc, hbv) and store in a list ordered as we expect.
+        t2w_scan_filename = scan_filenames[[i for i, sf in enumerate(scan_filenames) if "t2w" in sf][0]]
+        adc_scan_path_filename = scan_filenames[[i for i, sf in enumerate(scan_filenames) if "adc" in sf][0]]
+        hbv_scan_path_filename = scan_filenames[[i for i, sf in enumerate(scan_filenames) if "hbv" in sf][0]]
+        ordered_scan_filenames = [t2w_scan_filename, adc_scan_path_filename, hbv_scan_path_filename]
```

Comment:
Perhaps I'm misunderstanding the underlying structure of `scan_filenames`, but the filtrations here seem more complicated than needed. Perhaps the following would work and be a bit simpler?
```
t2w_scan_filename = [scan_filename in scan_filenames if "t2w" in scan_filename]
adc_scan_path_filename = [scan_filename in scan_filenames if "adc" in scan_filename]
hbv_scan_path_filename = [scan_filename in scan_filenames if "hbv" in scan_filename]
ordered_scan_filenames = t2w_scan_filename + adc_scan_path_filename + hbv_scan_path_filename
```
Do you ever expect there will be more than a single entry in these lists? The indexing [0] suggests maybe there could be more, and we only want one? You could do a similar index in the list filters above if that is the case and go back to `ordered_scan_filenames = [t2w_scan_filename, adc_scan_path_filename, hbv_scan_path_filename]`. You could also add asserts to make sure the resulting lists have length 1.

Reply:
Great suggestion!